src/streaming/array.writer.cpp

```diff
@@ -0,0 +1,422 @@
+
+#include <cmath>
+#include <functional>
+#include <latch>
+#include <stdexcept>
+
+#ifdef min
+#undef min
+#endif
+
+bool
+zarr::downsample(const ArrayWriterConfig& config,
+                 ArrayWriterConfig& downsampled_config)
+{
+    // downsample dimensions
+    std::vector<ZarrDimension> downsampled_dims;
+    for (auto i = 0; i < config.dimensions->ndims(); ++i) {
+        const auto& dim = config.dimensions->at(i);
+        // don't downsample channels
+        if (dim.type == ZarrDimensionType_Channel) {
+            downsampled_dims.push_back(dim);
+        } else {
+            const uint32_t array_size_px =
+              (dim.array_size_px + (dim.array_size_px % 2)) / 2;
+
+            const uint32_t chunk_size_px =
+              dim.array_size_px == 0
+                ? dim.chunk_size_px
+                : std::min(dim.chunk_size_px, array_size_px);
+
+            CHECK(chunk_size_px);
+            const uint32_t n_chunks =
+              (array_size_px + chunk_size_px - 1) / chunk_size_px;
+
+            const uint32_t shard_size_chunks =
+              dim.array_size_px == 0
+                ? 1
+                : std::min(n_chunks, dim.shard_size_chunks);
+
+            downsampled_dims.emplace_back(dim.name,
+                                          dim.type,
+                                          array_size_px,
+                                          chunk_size_px,
+                                          shard_size_chunks);
+        }
+    }
+    downsampled_config.dimensions = std::make_shared<ArrayDimensions>(
+      std::move(downsampled_dims), config.dtype);
+
+    downsampled_config.level_of_detail = config.level_of_detail + 1;
+    downsampled_config.bucket_name = config.bucket_name;
+    downsampled_config.store_path = config.store_path;
+
+    downsampled_config.dtype = config.dtype;
+
+    // copy the Blosc compression parameters
+    downsampled_config.compression_params = config.compression_params;
+
+    // can we downsample downsampled_config?
+    for (auto i = 0; i < config.dimensions->ndims(); ++i) {
+        // downsampling made the chunk size strictly smaller
+        const auto& dim = config.dimensions->at(i);
+        const auto& downsampled_dim = downsampled_config.dimensions->at(i);
+
+        if (dim.chunk_size_px > downsampled_dim.chunk_size_px) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/// Writer
+zarr::ArrayWriter::ArrayWriter(
+  const ArrayWriterConfig& config,
+  std::shared_ptr<ThreadPool> thread_pool,
+  std::shared_ptr<S3ConnectionPool> s3_connection_pool)
+  : config_{ config }
+  , thread_pool_{ thread_pool }
+  , s3_connection_pool_{ s3_connection_pool }
+  , bytes_to_flush_{ 0 }
+  , frames_written_{ 0 }
+  , append_chunk_index_{ 0 }
+  , is_finalizing_{ false }
+{
+}
+
+size_t
+zarr::ArrayWriter::write_frame(const uint8_t* data, size_t nbytes)
+{
+    const size_t nbytes_frame =
+      bytes_of_frame(*config_.dimensions, config_.dtype);
```

Comment:
Suggestion: Use `auto` more consistently. If you need a type conversion, make it explicit by using `static_cast`.